File: fasta/Version3/Gemini2.5/fasta3.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define LINE_LENGTH 60
#define IM 139968.0
#define IA 3877.0
#define IC 29573.0

// The ALU sequence
const char *alu =
    "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG"
    "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA"
    "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT"
    "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA"
    "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG"
    "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC"
    "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

// Struct to hold character-probability pairs
typedef struct {
    char c;
    double p;
} amino_acid;

// IUB ambiguity codes and probabilities
amino_acid iub[] = {
    {'a', 0.27}, {'c', 0.12}, {'g', 0.12}, {'t', 0.27},
    {'B', 0.02}, {'D', 0.02}, {'H', 0.02}, {'K', 0.02},
    {'M', 0.02}, {'N', 0.02}, {'R', 0.02}, {'S', 0.02},
    {'V', 0.02}, {'W', 0.02}, {'Y', 0.02}
};

// Homo sapiens frequencies
amino_acid homosapiens[] = {
    {'a', 0.3029549426680},
    {'c', 0.1979883004921},
    {'g', 0.1975473066391},
    {'t', 0.3015094502008}
};

// Creates cumulative probability tables for faster lookups
void make_cumulative(const amino_acid* table, int size, double* P, char* C) {
    double prob = 0.0;
    for (int i = 0; i < size; ++i) {
        prob += table[i].p;
        P[i] = prob;
        C[i] = table[i].c;
    }
    // Ensure the last probability is 1.0 for safety with floating point math
    P[size - 1] = 1.0;
}

// Repeats the source string `src` until `n` characters have been printed
void repeat_fasta(const char* src, int n) {
    int src_len = strlen(src);
    int pos = 0;
    char line_buffer[LINE_LENGTH + 1];
    line_buffer[LINE_LENGTH] = '\n';

    int full_lines = n / LINE_LENGTH;
    for (int i = 0; i < full_lines; ++i) {
        for (int j = 0; j < LINE_LENGTH; ++j) {
            line_buffer[j] = src[pos];
            pos = (pos + 1) % src_len;
        }
        fwrite(line_buffer, 1, LINE_LENGTH + 1, stdout);
    }

    int remaining_chars = n % LINE_LENGTH;
    if (remaining_chars > 0) {
        for (int i = 0; i < remaining_chars; ++i) {
            line_buffer[i] = src[pos];
            pos = (pos + 1) % src_len;
        }
        line_buffer[remaining_chars] = '\n';
        fwrite(line_buffer, 1, remaining_chars + 1, stdout);
    }
}

// Generates and prints a random sequence based on the provided probability table
double random_fasta(const amino_acid* table, int size, int n, double seed) {
    double P[size];
    char C[size];
    make_cumulative(table, size, P, C);

    char line_buffer[LINE_LENGTH + 1];
    line_buffer[LINE_LENGTH] = '\n';

    int full_lines = n / LINE_LENGTH;
    for (int i = 0; i < full_lines; ++i) {
        for (int j = 0; j < LINE_LENGTH; ++j) {
            seed = fmod(seed * IA + IC, IM);
            double r = seed / IM;
            
            // Find character by checking against cumulative probabilities
            // This is equivalent to Python's bisect.bisect
            char next_char;
            for (int k = 0; k < size; ++k) {
                if (r < P[k]) {
                    next_char = C[k];
                    break;
                }
            }
            line_buffer[j] = next_char;
        }
        fwrite(line_buffer, 1, LINE_LENGTH + 1, stdout);
    }

    int remaining_chars = n % LINE_LENGTH;
    if (remaining_chars > 0) {
        for (int i = 0; i < remaining_chars; ++i) {
            seed = fmod(seed * IA + IC, IM);
            double r = seed / IM;
            
            char next_char;
            for (int k = 0; k < size; ++k) {
                if (r < P[k]) {
                    next_char = C[k];
                    break;
                }
            }
            line_buffer[i] = next_char;
        }
        line_buffer[remaining_chars] = '\n';
        fwrite(line_buffer, 1, remaining_chars + 1, stdout);
    }
    
    return seed;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <n>\n", argv[0]);
        return 1;
    }
    int n = atoi(argv[1]);

    fputs(">ONE Homo sapiens alu\n", stdout);
    repeat_fasta(alu, n * 2);

    fputs(">TWO IUB ambiguity codes\n", stdout);
    // The initial seed is 42.0, as in the Python script
    double seed = 42.0; 
    seed = random_fasta(iub, sizeof(iub)/sizeof(amino_acid), n * 3, seed);

    fputs(">THREE Homo sapiens frequency\n", stdout);
    random_fasta(homosapiens, sizeof(homosapiens)/sizeof(amino_acid), n * 5, seed);

    return 0;
}

File: fasta/Version3/Gemini2.5/fasta3.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// translated from Python to Rust by a large language model

use std::io::{stdout, Write, BufWriter};
use std::cmp::min;

const LINE_WIDTH: usize = 60;

const ALU: &str = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG\
                   GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA\
                   CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT\
                   ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA\
                   GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG\
                   AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC\
                   AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

// IUB ambiguity codes and their probabilities
const IUB: &[(u8, f64)] = &[
    (b'a', 0.27), (b'c', 0.12), (b'g', 0.12), (b't', 0.27),
    (b'B', 0.02), (b'D', 0.02), (b'H', 0.02), (b'K', 0.02),
    (b'M', 0.02), (b'N', 0.02), (b'R', 0.02), (b'S', 0.02),
    (b'W', 0.02), (b'Y', 0.02), (b'V', 0.02),
];

// Homo sapiens DNA frequencies
const HOMOSAPIENS: &[(u8, f64)] = &[
    (b'a', 0.3029549426680),
    (b'c', 0.1979883004921),
    (b'g', 0.1975473066391),
    (b't', 0.3015094502008),
];

/// From a table of characters and their probabilities, generate cumulative probabilities.
/// This is used for fast random selection.
fn make_cumulative(table: &[(u8, f64)]) -> (Vec<f64>, Vec<u8>) {
    let mut probs = Vec::with_capacity(table.len());
    let mut chars = Vec::with_capacity(table.len());
    let mut p = 0.0;
    for &(char, prob) in table {
        p += prob;
        probs.push(p);
        chars.push(char);
    }
    // Ensure the last probability is exactly 1.0 to handle potential floating point inaccuracies
    if let Some(last_prob) = probs.last_mut() {
        *last_prob = 1.0;
    }
    (probs, chars)
}

/// Repeats the `src` sequence to a total length of `n`, printing it in lines of `LINE_WIDTH`.
fn repeat_fasta(src: &[u8], n: usize, writer: &mut impl Write) -> std::io::Result<()> {
    let mut output_bytes = Vec::with_capacity(n + n / LINE_WIDTH);
    output_bytes.extend(src.iter().cycle().take(n));
    
    for chunk in output_bytes.chunks(LINE_WIDTH) {
        writer.write_all(chunk)?;
        writer.write_all(b"\n")?;
    }
    Ok(())
}

// Constants for the Linear Congruential Generator (LCG)
const IM: f64 = 139968.0;
const IA: f64 = 3877.0;
const IC: f64 = 29573.0;
static mut SEED: f64 = 42.0;

/// Generates a random DNA sequence based on the provided probability table.
fn random_fasta(table: &[(u8, f64)], n: usize, writer: &mut impl Write) -> std::io::Result<()> {
    let (probs, chars) = make_cumulative(table);
    let mut line_buf = [0u8; LINE_WIDTH];
    
    let mut remaining = n;
    while remaining > 0 {
        let line_len = min(remaining, LINE_WIDTH);

        for i in 0..line_len {
            // This is unsafe because SEED is a mutable static, which is not thread-safe.
            // For this single-threaded benchmark, it's a direct translation of the Python logic.
            unsafe {
                SEED = (SEED * IA + IC) % IM;
                let r = SEED / IM;
                // Find the first cumulative probability that is >= our random number
                let index = probs.iter().position(|&p| p >= r).unwrap();
                line_buf[i] = chars[index];
            }
        }

        writer.write_all(&line_buf[..line_len])?;
        writer.write_all(b"\n")?;
        remaining -= line_len;
    }
    Ok(())
}

fn main() -> std::io::Result<()> {
    let n = std::env::args_os()
        .nth(1)
        .and_then(|s| s.into_string().ok())
        .and_then(|s| s.parse().ok())
        .unwrap_or(1000);

    let mut writer = BufWriter::new(stdout().lock());

    writer.write_all(b">ONE Homo sapiens alu\n")?;
    repeat_fasta(ALU.as_bytes(), n * 2, &mut writer)?;

    writer.write_all(b">TWO IUB ambiguity codes\n")?;
    random_fasta(&IUB, n * 3, &mut writer)?;

    writer.write_all(b">THREE Homo sapiens frequency\n")?;
    random_fasta(&HOMOSAPIENS, n * 5, &mut writer)?;
    
    writer.flush()?;
    Ok(())
}
